Add Python script for automated bindings

zer0main opened the pull request.

zer0main committed:
Add blank file with small description of script

zer0main committed:
Some simple functions for generating function names

Just tasting Python.

zer0main committed:
Add getComplexArgument() function for code generation

zer0main committed:
Add returnComplexValue() function

zer0main committed:
Add callWtFunction()

zer0main committed:
returnComplexValue(): name of return value is result

zer0main committed:
Add Dictionary TYPE_FROM_LUA_FUNCS

It's just a small blank for future type conversion system.

zer0main committed:
Add Dictionary TYPE_TO_LUA_FUNCS

zer0main committed:
getComplexArgument(): use one dict arg instead of multiple args

zer0main committed:
Add func generateIncludes()

zer0main committed:
Add function getInbuiltTypeArgument()

zer0main committed:
Function returnInbuiltTypeValue()

zer0main committed:
callWtFunction(); args: list of strings ---> list of objects

zer0main committed:
Automate bindings: add isInbult() func

zer0main committed:
Automate bindings: add implementLuaCFunction() func

zer0main committed:
Automate bindings: rename some funcs to use the same style of naming

zer0main committed:
Automate bindings: add generateMethodsArray() func

zer0main committed:
Automate bindings: add generateModuleFunc() func

zer0main committed:
Add generateModule() function

zer0main committed:
Automate bindings: add parse() function

Use pygccxml for reading output of GCCXML and for running it.

zer0main committed:
Automate bindings: remake generateSkeleton() to writeSourceToFile()

zer0main committed:
Remove useless code

zer0main committed:
Automate bindings: add getMethods() function

zer0main committed:
Rewrite bind() function

zer0main committed:
Rename parser() to parse()

The second name is more suitable because functions are
actions so verbs are better to describe them.

zer0main committed:
automate_bindings.py: do not use Python3

pygccxml is not ported to Python3.

zer0main committed:
automate_bindings.py: fix several typos

zer0main committed:
automate_bindings.py: fix writeSourceToFile() [trigger]

zer0main committed:
Automate bindings: fix strings formatting with dictionaries

Add missing 's' after %(expression).

zer0main committed:
automate_bindings.py: use member_functions() in getMethods()

zer0main committed:
Automate bindings: fix strings formatting

s % a % b ----> s % (a, b)

zer0main committed:
Automate bindings: method doesn't have field args

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,207 @@
+
+from pygccxml import utils
+from pygccxml import declarations
+from pygccxml import parser
```
sort deps

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,207 @@
+
+    # Get access to the global namespace
+    global_namespace = declarations.get_global_namespace(decls)
+    return global_namespace
```
remove empty lines from this function

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,207 @@
+
+def getMethods(global_namespace):
+    Wt = global_namespace.namespace("Wt")
+    base = ''
```
base = Nona

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,207 @@
+    mathods = []
+    for decl in Wt.declarations:
+        if isinstance(decl, declarations.class_declaration_t):
+            base = str(decl.bases[0])
```
Add `assert base is None` before the assignment.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,207 @@
+    if return_type == 'void':
+        return func_s
+    else:
+        return '%s result = %s' % return_type % func_s
```
% (return_type, func_s)

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,207 @@
+
+def implementLuaCFunction(module_name, method_name, args, return_type):
+    head = 'int luawt_%s_%s(lua_State* L) {' % module_name % method_name
+    body = ''
```
Replace with list to reduce number of copies. O(N^2) -> O(N)

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,207 @@
+        else:
+            body += getComplexArgument(options)
+    body += callWtFunction(return_type, args, method_name)
+    if (isInbuilt(return_type)):
```
remove extra parenthesis

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,207 @@
+    return code % options
+
+def generateModule(module_name, methods, base):
+    source = ''
```
Replace with list to reduce number of copies. O(N^2) -> O(N)

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,207 @@
+            module_name,
+            method.name,
+            method.args,
+            method.return_type
```
add `,`

zer0main committed:
Fix serveral mistakes; make a small refactoring

Mistakes:
erase arg.index - use arg_n instead;
Convert arg.decl_type and return_type to strings before
passing it to other functions.

Refactoring:
isInbuilt() --> getInbuiltType();
unite returnInbuiltTypeValue() and returnComplexValue()

zer0main committed:
Sort deps

zer0main committed:
Improve the style

zer0main committed:
Replace '' by None

zer0main committed:
Add `assert base is None` before the assignment

zer0main committed:
automate_bindings.py: fix indent problems

zer0main committed:
Automate bindings: add 2 to arg_n

In Lua indices start with 1; the first one is object itself,
so we have to add 2 to arg_n.

zer0main committed:
automate_bindings.py: fix indent problems in getArgument() funcs

zer0main committed:
automate_bindings.py: add getSelf() function and use it

zer0main committed:
automate_bindings.py: do not use a loop to get bases

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,215 @@
+            'index' : arg_n + 2,
+        }
+        arg_type = getInbuiltType(str(arg.decl_type))
+        if (arg_type):
```
Remove extra parenthesis.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,215 @@
+    body = getSelf(module_name)
+    arg_n = 0
+    while arg_n < len(args):
+        arg = args[arg_n]
```
``` python
for i, arg in enumarate(args):
```

And replace `arg_n` with `i`.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,215 @@
+    return_type = getInbuiltType(str(return_type))
+    body += returnValue(return_type)
+    close = '}\n\n'
+    return head + body + close
```
Remove `head` and `close` and put everything in one format string:

``` python
return r'''int luawt_%(module)s_%(method)s(lua_State* L) {
  %(body)s
}
''' % {
  'module': module_name,
  'method': method_name,
  'body': body,
}
```

And do the same in `generateMethodsArray`.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,215 @@
+TYPE_TO_LUA_FUNCS = {
+    'int' : 'lua_pushinteger',
+    'bool' : 'lua_pushboolean',
+}
```
``` python
TYPE_TO_LUA_FUNCS = {
    value: key
    for (key, value)
    in TYPE_FROM_LUA_FUNCS.items()
}
```

zer0main replied:
I'm sorry, but could you please explain this syntactic sugar to me? If I understand it correctly, it simply creates copy of `TYPE_FROM_LUA_FUNCS`. But they must not be equal: similar keys but different values (functions names in our case).

starius replied:
Sorry, I overlooked that values are different. Ignore this comment.

zer0main replied:
You're expected to mark this change request as closed.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,215 @@
+RETURN_CALLS_TEMPLATE = r'''
+    %s(L, result);
+    return 1;
+'''
```
`INCLUDES_TEMPLATE`, `MODULE_FUNC_TEMPLATE` and `RETURN_CALLS_TEMPLATE` are used only once. Can you inline them?

zer0main replied:
Nope, because of several indents problems.

starius replied:
At least define them closer to where they are used. Otherwise they look like something used by multiple functions or even other modules.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,215 @@
+
+def getInbuiltType(type):
+    for inbuilt_type in TYPE_FROM_LUA_FUNCS:
+        if type.find(inbuilt_type) is not -1:
```
``` python
if inbuilt_type not in type:
```

Can you also rename `type` to something else, because `type` is a built-in function in Python, which returns type of its argument?

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,215 @@
+def getMethods(global_namespace):
+    Wt = global_namespace.namespace("Wt")
+    methods = Wt.member_functions()
+    main_class = (Wt.classes())[0]
```
Remove extra parenthesis.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,215 @@
+    return global_namespace
+
+def getMethods(global_namespace):
+    Wt = global_namespace.namespace("Wt")
```
Replace `"` with `'`, because you use `'` for other strings.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,215 @@
+
+from pygccxml import declarations
+from pygccxml import parser
+from pygccxml import utils
```
Can you `import pygccxml` and use full names (e.g. `pygccxml.parser`)?

Name `parser` is reused in `main` function for arguments parser.

zer0main committed:
automate_bindings.py: import pygccxml and use full names

E.g. parser -> pygccxml.parser
Name parser is reused in main function for arguments parser.

zer0main committed:
Move some template strings closer to where they are used

Otherwise they look like something used by multiple functions
or even other modules.

zer0main committed:
Replace " with ', because ' is used for other strings

zer0main committed:
automate_bindings.py: remove extra parenthesis

zer0main committed:
getInbuiltType(): rename type to full_type

`type` is a built-in function in Python, which returns
type of its argument.

zer0main committed:
Use enumerate() in implementLuaCFunction()

zer0main committed:
generateMethodsArray(), implmentLuaCFunction(): use single format string

Erase `head` and `close` and put everything in one format string.

zer0main committed:
generateModuleFunc(): use lstrip() to fix indent problems

zer0main committed:
automate_bindings.py: replace some strings by lists

To reduce number of copies. O(N^2) -> O(N)

zer0main committed:
METHODS_ARRAY_TEMPLATE: add missing ; at the end

zer0main committed:
automate_bindings.py: pass 0 to DECLARE_CLASS if there is no base

Rename getMethods() to getMethodsAbdBases().
Generate comment in case of base absence.

zer0main committed:
returnValue(): fix indent problem in case of `void` type

zer0main committed:
automate_bindings.py: do not convert None to str

Because 'None' is True.

zer0main committed:
Automate bindings: add functions for make() generation

implementLuaCFunction() was slightly changed to make it more universal.
Now it works for both methods and constructors.

zer0main committed:
Use WObject as base in case of base absence

Classes that are not descendants of WObject do not exist
as classes for our system.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,270 @@
+    # Configure the xml generator
+    xml_generator_config = pygccxml.parser.xml_generator_configuration_t(
+        xml_generator_path = generator_path,
+        xml_generator = generator_name
```
Comma

zer0main committed:
Add missing comma

zer0main committed:
Clarify some comments

zer0main committed:
callWtConstructor(): add correct implementation

zer0main committed:
implementLuaCFunction(): fix args' indices offset

In case of constructor offset must be equal to 1.

zer0main committed:
Add function writeToFile()

zer0main committed:
Use writeToFile() func in writeSourceToFile()

zer0main committed:
Add function readFile()

zer0main committed:
automate_bindings.py: add addItem() function

zer0main committed:
Automate bindings: addItemToFiles() function

zer0main committed:
Add addModuleToLists() function; call it from bind()

zer0main committed:
Travis: add comment for future TODO about automate_bindings' tests

zer0main committed:
Another TODO for zer0main

zer0main commented:
The work was a little bit greater than I expect so all further improvements and development should be moved to the next pull request. I'm going to merge this one.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,321 @@
+        list_beginning += 1
+    curr_index = list_beginning
+    # TODO (for zer0main).
+    # We need to consider that base must be before child in init.cpp
```
s/consider/take into account/

zer0main replied:
Done :+1:

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,321 @@
+def writeSourceToFile(module_name, source):
+    writeToFile('src/luawt/%s.cpp' % module_name, source)
+
+def addItemToFiles(filenames, patterns, module_strs):
```
Can you replace three lists with a list of dicts?

zer0main replied:
done :+1:

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,321 @@
+    return frame.lstrip() % options
+
+def getComplexArgument(options):
+    frame = r'''
```
``` python
return r'''%(argument_type)s* %(argument_name)s =
        luawt_checkFromLua<%(argument_type)s>(L, %(index)s);
''' % options
```

zer0main replied:
No. I won't change this code because it  was utterly hard to fix all the indent problems in generated code. lstrip(), indents and stuff must stay the same.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,321 @@
+            func_name = 'luawt_toLua'
+        return RETURN_CALLS_TEMPLATE % func_name
+
+def getInbuiltType(full_type):
```
Maybe s/inbuilt/built-in/ ?

zer0main replied:
There is no word builtin (without dash) in English. Dash is totally unacceptable for style reasons.
The word `inbuilt` exists is and it's quite good for this situation.

starius replied:
Builtin is acceptable and more commonly used in code.

Compare:
https://github.com/search?utf8=%E2%9C%93&q=builtin&type=Code
https://github.com/search?utf8=%E2%9C%93&q=inbuilt&type=Code

zer0main replied:
Ok, done.

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,321 @@
+def addItem(pattern, added_str, content):
+    list_beginning = 0
+    while not re.search(pattern, content[list_beginning]):
+        list_beginning += 1
```
Maybe move three lines above to a function:

``` python
def foo():
    for i, text in enumerate(content):
        if re.search(pattern, text):
            return i
```

?

zer0main replied:
Done :+1:

starius commented on tools/automate_bindings.py:
```diff
@@ -0,0 +1,321 @@
+        f.write(what)
+
+def readFile(filename):
+    with open(filename, 'r') as f:
```
Why 'wt' in writeToFile, but 'r' (not 'rt') here?

zer0main replied:
Done :+1:

zer0main committed:
readFile(): open file with 'rt' flag

't' means text file.

zer0main committed:
Clarify some comments

zer0main committed:
automate_bindings.py: add getIndexOfFirstMatch()

Use it in addItem().
Code seems clearer.

zer0main committed:
addModuleToLists(): replace three lists with a list of dicts

zer0main committed:
automate_bindings.py: s/inbuilt/builtin/
